pimd/pim_pim.c

```diff
@@ -585,14 +585,70 @@ static int pim_msg_send_frame(int fd, char *buf, size_t len,
 		return pim_msg_send_frame(fd, (char *)ip2, sendlen, dst, salen,
 					  ifname);
 	}
-#endif
 
 	zlog_warn(
 		"%s: sendto() failure to %pSU: iface=%s fd=%d msg_size=%zd: %m",
 		__func__, dst, ifname, fd, len);
 	return -1;
 }
 
+#else
+static int pim_msg_send_frame(pim_addr src, pim_addr dst, ifindex_t ifindex,
+			      struct iovec *message, int fd)
+{
+	int retval;
+	struct msghdr smsghdr;
+	struct cmsghdr *scmsgp;
+	union {
+		struct cmsghdr hdr;
+		uint8_t buf[CMSG_SPACE(sizeof(struct in6_pktinfo))];
+	} cmsgbuf;
+	struct in6_pktinfo *pktinfo;
+	struct sockaddr_in6 dst_sin6;
+
+	memset(&cmsgbuf, 0, sizeof(cmsgbuf));
+	scmsgp = (struct cmsghdr *)&cmsgbuf;
+	pktinfo = (struct in6_pktinfo *)(CMSG_DATA(scmsgp));
+	memset(&dst_sin6, 0, sizeof(struct sockaddr_in6));
+
+	/* source address */
+	pktinfo->ipi6_ifindex = ifindex;
+	pktinfo->ipi6_addr = src;
+
+	/* destination address */
+	dst_sin6.sin6_family = AF_INET6;
+#ifdef SIN6_LEN
+	dst_sin6.sin6_len = sizeof(struct sockaddr_in6);
+#endif /*SIN6_LEN*/
+	dst_sin6.sin6_addr = dst;
+	dst_sin6.sin6_scope_id = ifindex;
+
+
+	/* send msg hdr */
+	memset(&smsghdr, 0, sizeof(smsghdr));
+	smsghdr.msg_iov = message;
+	smsghdr.msg_iovlen = 1;
+	smsghdr.msg_name = (caddr_t)&dst_sin6;
+	smsghdr.msg_namelen = sizeof(dst_sin6);
+	smsghdr.msg_control = (caddr_t)&cmsgbuf.buf;
+	smsghdr.msg_controllen = CMSG_LEN(sizeof(cmsgbuf.buf));
+	scmsgp = CMSG_FIRSTHDR(&smsghdr);
+
+	scmsgp->cmsg_level = IPPROTO_IPV6;
```

Comment:
…here?

The PKTINFO related lines should be properly organised next to each other.

Reply:
i have regrouped the related statements. can you please check

Reply:
yes, only the `scpmsg` line above is now useless